### ISSUE TITLE
pack/ee.blockUpacker.Unpack: checksum mismatch

### ISSUE THREAD

**Issue by masiulaniec** [upspin/upspin#415]
My upspin installation was working fine unattended until suddenly all clients broke (both upspin cli and upspinfs). Here's the access error in detail:

```
$ upspin ls @/
upspin: ls: client.Lookup:
        dir/remote.Lookup("remote,mlnc.me:443"):
        dir/server.Lookup:
        dir/server/tree.Lookup:
        jacek.masiulaniec@gmail.com/: client/clientutil.ReadAll:
        pack/ee.blockUpacker.Unpack: checksum mismatch
$
```

Any ideas on how to diagnose this error?

**Comment by masiulaniec** [comment]
I'm running the following code:

```
commit a9dd198264d4ebf3adce05602c8f1c5ed68df3d6
Date:   Thu Apr 27 15:43:54 2017 -0700
```

It dates back to April so it's pretty old. I will refrain from upgrading until somebody more knowledgable suggests it might be a good idea.

**Comment by masiulaniec** [comment]
Restoring the `dirserver-logs` directory from a backup seems to have resolved the problem.

If I didn't make backups daily, would that mean loss of access to my data?

**Comment by adg** [comment]
> If I didn't make backups daily, would that mean loss of access to my data?

Potentially, yes. 😦 I filed issue #419 to track this.

> My upspin installation was working fine unattended until suddenly all clients broke

Did you see any error messages in the server logs? It looks to me like a blob on the storeserver got corrupted somehow, and perhaps by restoring from a backup the dirserver was no longer trying to read that particular blob. It's strange that the blob got corrupted, though. Are you storing them in a cloud storage service, or on local disk?

At the very least we should make the error message report the blob that was trying to be unpacked when the error occurred. That'd be a change to `client/clientutil.ReadAll`.

**Comment by adg** [comment]
cc @edpin: does my analysis sound correct to you?

**Comment by masiulaniec** [comment]
> Are you storing them in a cloud storage service, or on local disk?

I'm using GCS.

> Did you see any error messages in the server logs?

I found some GSCS errors, all falling into two classes:

```
cloud/storage/gcs: WARNING: retrying Insert(...): googleapi: got HTTP response code 503 with body: Service Unavailable
cloud/storage/gcs: WARNING: retrying Insert(...): googleapi: got HTTP response code 503 with body:
```

Where `...` stands for some hashes. Note that the second line is not a typo, the http body was probably an empty string.

**Comment by adg** [comment]
Oh no! This is a bad bug. It turns out that in the case of a retry, the next attempted Put would always put an empty object. 😮 

I sent a fix: https://upspin-review.googlesource.com/10642

**Comment by adg** [comment]
The fix is now committed: 909d8bb62fdfa7b2e27393bc4d861e7f4939c3a3

Thanks for the report, and sorry for the inconvenience. Please follow issue #419 to track the resolution of the underlying concern, that of restoring a tree after loss of local disk.
